common/libraries/clients/postgresql/src/main/java/org/eclipse/dataspaceconnector/clients/postgresql/PostgresqlClient.java

```diff
@@ -0,0 +1,72 @@
+/*
+ *  Copyright (c) 2021 Daimler TSS GmbH
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Daimler TSS GmbH - Initial API and Implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.clients.postgresql;
+
+import org.eclipse.dataspaceconnector.clients.postgresql.row.RowMapper;
+
+import java.sql.SQLException;
+import java.util.List;
+
+public interface PostgresqlClient extends AutoCloseable {
+
+    /**
+     * Intended for mutating queries.
+     *
+     * @param sql       the parametrized sql query
+     * @param arguments the parameteres to interpolate with the parametrized sql query
+     * @return rowsChanged
+     * @throws SQLException if execution of the query was failing
+     */
+    int execute(String sql, Object... arguments) throws SQLException;
+
+    /**
+     * Intended for reading queries.
+     *
+     * @param rowMapper able to map a row to an object e.g. pojo.
+     * @param sql       the parametrized sql query
+     * @param arguments the parameteres to interpolate with the parametrized sql query
+     * @param <T>       generic type returned after mapping from the executed query
+     * @return results
+     * @throws SQLException if execution of the query or mapping was failing
+     */
+    <T> List<T> execute(RowMapper<T> rowMapper, String sql, Object... arguments) throws SQLException;
+
+    /**
+     * Spans a transaction around the passed callback.
+     *
+     * @param callback transaction scoped callback
+     * @throws SQLException if the transaction was failing
+     */
+    void doInTransaction(TransactionCallback callback) throws SQLException;
```

Comment:
I'm not into callbacks too, I think they made the code harder to read/test.
The alternative could be to have a method (like `transaction`, or `begin`) that returns a TransactionContext that would take care of handling the transaction.

Reply:
The problem with your proposal is, that one could easily leak a connection if not properly returned to the pool, which then means the connection is blocked for all other accessors - So think about it as it would be a security net. 
One isn't forced to use the PostgresqlClient implementation - it would be able to obtain a `Connection` from a `ConnectionFactory`/`ConnectionPool` on its own and roll his own transaction handling.